mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+import itertools
+import random
+from collections.abc import Iterable
+from functools import cache, cached_property
+from typing import Callable, Optional
+
+from .. import Agent
+
+Coordinate = tuple[int, int]
+
+
+class Cell:
+    __slots__ = ["coordinate", "_connections", "agents", "capacity", "properties"]
+
+    def __init__(self, i: int, j: int, capacity: int | None = 1) -> None:
+        self.coordinate = (i, j)
+        self._connections: list[Cell] = []
+        self.agents: list[Agent] = []
+        self.capacity = capacity
+        self.properties: dict[str, object] = {}
+
+    def connect(self, other) -> None:
+        """Connects this cell to another cell."""
+        self._connections.append(other)
+
+    def disconnect(self, other) -> None:
+        """Disconnects this cell from another cell."""
+        self._connections.remove(other)
+
+    def add_agent(self, agent: Agent) -> None:
+        """Adds an agent to the cell."""
+        if self.capacity and len(self.agents) >= self.capacity:
+            raise Exception("ERROR: Cell is full")
+        if isinstance(agent.cell, Cell):
+            agent.cell.remove_agent(agent)
+        self.agents.append(agent)
+        agent.cell = self
+
+    def remove_agent(self, agent: Agent) -> None:
+        """Removes an agent from the cell."""
+        self.agents.remove(agent)
+        agent.cell = None
+
+    @property
+    def is_empty(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
+    def neighborhood(self, radius=1, include_center=False):
+        if radius == 0:
+            return {self: self.agents}
+        if radius == 1:
+            return CellCollection(
+                {neighbor: neighbor.agents for neighbor in self._connections}
+            )
+        else:
+            neighborhood = {}
+            for neighbor in self._connections:
+                neighborhood.update(neighbor.neighorhood(radius - 1, include_center))
+            if not include_center:
+                neighborhood.pop(self, None)
+            return CellCollection(neighborhood)
+
+
+class CellCollection:
```

Comment:
We now have AgentSet and CellCollection. Mixed feelings about CellSet.

Reply:
Can't be CellSet because the elements in this object is ordered. Some agents might want to actually loop the Cell elements according to the given non-random order, e.g. a bee that scans in clockwise/counterclockwise direction starting from the north direction.

Reply:
The current draft is also implicitly immutable because of the `@cached_property`.

The AgentSet is also ordered by the way, so that is not really a good reason in my view. 

The direction is interesting but we currently don't explicitly store this information in cell.connections. In my proof of principle, I was storing everything separately [topleft, top., toplight, left, etc.]. The current code is much more concise, cleaner, and it would be rare to need the directions. However, I am still wondering whether connections should not be a dict with the direction as key and the cell as value. You can do stupid stuff like

```python
diagonal = []
cell = self.cell.connections[l"top left"] 
while cell is not self:
    diagonal.append(cell)
    cell = cell.connections["top left"] 
```

Which would give you the diagonal cells on a torus space. This is an example of how the current data structure is like a linked list.